### ISSUE TITLE
`prometheus_client` mounts metrics to `/metrics/` instead of `/metrics`

### ISSUE THREAD

**Issue by hmellor** [prometheus/client_python#1016]
The [documentation](https://prometheus.github.io/client_python/exporting/http/fastapi-gunicorn/) gives an example application which mounts the ASGI app to `/metrics`

```python
from fastapi import FastAPI
from prometheus_client import make_asgi_app

# Create app
app = FastAPI(debug=False)

# Add prometheus asgi middleware to route /metrics requests
metrics_app = make_asgi_app()
app.mount("/metrics", metrics_app)
```

and says that you should be able to see the metrics at http://localhost:8000/metrics, however if you try to access this endpoint

```console
$ uvicorn app:app
INFO:     Started server process [1620866]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     127.0.0.1:34428 - "GET / HTTP/1.1" 404 Not Found
INFO:     127.0.0.1:34428 - "GET /favicon.ico HTTP/1.1" 404 Not Found
INFO:     127.0.0.1:34444 - "GET /metrics HTTP/1.1" 307 Temporary Redirect
INFO:     127.0.0.1:34444 - "GET /metrics/ HTTP/1.1" 200 OK
```

you get redirected to `/metrics/` which, from my experience, is incompatible with the default `metrics_path` in [<scrape_config>](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config).

I know I could just set `metrics_path: /metrics/`, but that feels like the wrong solution.

**Comment by csmarchbanks** [comment]
@kakkoyun this might be something to look at to learn more about the project as well.

**Comment by Sakoes** [comment]
Thanks for pointing this out @hmellor. I was am doing something similar but got an "Method Not Allowed" when doing a GET request for http://localhost:8000/metrics.
However, http://localhost:8000/metrics/ does seem to work

**Comment by Sakoes** [comment]
I think I found the root cause.
It has to do with how FastApi and Starlette (which FastApi is based on) are mounting sub-applications. A metrics sub-application is created ass follows:
```
metrics_app = make_asgi_app()
app.mount("/metrics", metrics_app)
```
And from FastApi and Starlette docs you can read that sub-applications are mounted under a prefix, in this case "/metrics". So when you want to actually access this sub-application I guess a new root endpoint is automatically added thus creating  "/metrics/".
Not sure if or how you could fix this, but it looks like this is intended behavior when running the prometheus client as a FastApi sub-application.

Here you can find the docs:
- [FastApi sub-applications](https://fastapi.tiangolo.com/advanced/sub-applications/#technical-details-root_path)
- [Starlette submounting routes](https://www.starlette.io/routing/#submounting-routes)

**Comment by hmellor** [comment]
In `aioprometheus` (a third party Prometheus client) they use `add_route` to expose the metrics and they don't have this problem

```python
from aioprometheus.asgi.starlette import metrics
app.add_route("/metrics", metrics)
```

**Comment by Sakoes** [comment]
Yes, that is exactly the cause. In this client the make_asgi_app() creates a _sub-application_ that is _mounted_ with _app.mount()_ **thus creating a "/metrics" prefix vs an actual route**.
While the aioprometheus.asgi.starlette implementation is adding a route directly with _app.add_route()_. So "/metrics" is the actual route and not a prefix.

**Comment by hmellor** [comment]
This workaround appears to work:

```python
import re
from starlette.routing import Mount

...

# Add prometheus asgi middleware to route /metrics requests
mount = Mount("/metrics", make_asgi_app())
mount.path_regex = re.compile('^/metrics(?P<path>.*)$')
app.routes.append(mount)
```

**Comment by hmellor** [comment]
@csmarchbanks @kakkoyun has this been fixed yet?
